File: 01/go/01.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func main() {
	// Load data from file
	data, err := ioutil.ReadFile("../01.txt")
	check(err)
	lines := strings.Split(string(data), "\n")
	nums := make([]int, len(lines)-1)
	for i, line := range lines {
		if len(line) == 0 {
			break
		}
		num, err := strconv.Atoi(line)
		check(err)
		nums[i] = num
	}

	// Sum frequency changes
	result := 0
	for _, num := range nums {
		result += num
	}

	// Output part 1 result
	fmt.Println(result)

	// Loop forever until same frequency observed twice
	m := make(map[int]int)
	result = 0
	i := 0
	for {
		idx := i % len(nums)
		result += nums[idx]
		_, ok := m[result]
		if ok {
			break
		} else {
			m[result] = result
		}
		i++
	}

	// Output part 2 result
	fmt.Println(result)
}

File: 01/python/01.py
#!/usr/bin/env python3
import numpy as np


if __name__ == '__main__':
    # Load data from file
    nums = np.loadtxt('../01.txt')

    # Output part 1 result
    print(int(np.sum(nums)))

    def find_repeat(nums):
        csum = np.zeros(len(nums))
        m = {}
        # Loop forever until same frequency observed twice
        while True:
            csum = csum[-1] + np.cumsum(nums)
            for n in csum:
                if n in m:
                    return n
                else:
                    m[n] = None

    # Output part 2 result
    print(int(find_repeat(nums)))
